solver/src/main/java/org/chocosolver/solver/constraints/nary/binPacking/PropBinPacking.java

```diff
@@ -0,0 +1,409 @@
+        this.itemBin = itemBin;
+        this.itemSize = itemSize;
+        this.binLoad = binLoad;
+        this.offset = offset;
+        this.useNoSumFiltering = useNoSumFiltering;
+
+        nbItems = itemBin.length;
+        nbAvailableBins = binLoad.length;
+
+        monitors = new IIntDeltaMonitor[nbItems];
+        for(int i=0; i<nbItems; i++){
+            monitors[i] = itemBin[i].monitorDelta(this);
+        }
+
+        P = new ISet[nbAvailableBins];
+        R = new ISet[nbAvailableBins];
+        sumR = new IStateInt[nbAvailableBins];
+        sumP = new IStateInt[nbAvailableBins];
+        for(int j = 0; j<nbAvailableBins; j++) {
+            P[j] = SetFactory.makeStoredSet(SetType.BITSET, 0, itemBin[0].getModel());
+            R[j] = SetFactory.makeStoredSet(SetType.BITSET, 0, itemBin[0].getModel());
+            int pj = 0;
+            int rj = 0;
+            for(int i = 0; i<nbItems; i++) {
+                if(itemBin[i].contains(j+offset)) {
+                    P[j].add(i);
+                    pj += itemSize[i];
+                    if(itemBin[i].isInstantiated()) {
+                        R[j].add(i);
+                        rj += itemSize[i];
+                    }
+                }
+            }
+            sumR[j] = itemBin[0].getModel().getEnvironment().makeInt(rj);
+            sumP[j] = itemBin[0].getModel().getEnvironment().makeInt(pj);
+        }
+
+        binsToProcess = new BitSet(nbAvailableBins);
+
+        // NoSum init
+        X = new int[itemSize.length];
+        indexSortedBySize = IntStream.range(0, itemSize.length)
+            .boxed()
+            .sorted(Comparator.comparingInt(i -> -itemSize[i]))
+            .mapToInt(i -> i)
+            .toArray();
+    }
+
+    @Override
+    public int getPropagationConditions(int vIdx) {
+        if(vIdx < nbItems) {
+            return IntEventType.all();
+        } else {
+            return IntEventType.boundAndInst();
+        }
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////
+    //////////////////////////////    FILTERING ALGORITHMS    //////////////////////////////
+    ////////////////////////////////////////////////////////////////////////////////////////
+
+    private void removeItemFromBin(int binIdx, int itemIdx) throws ContradictionException {
+        if(P[binIdx].contains(itemIdx)) {
+            P[binIdx].remove(itemIdx);
+            sumP[binIdx].add(-itemSize[itemIdx]);
+            binLoad[binIdx].updateUpperBound(sumP[binIdx].get(), this);
+            binsToProcess.set(binIdx);
+            if(itemBin[itemIdx].isInstantiated()) {
+                updateRAfterInstantiation(itemBin[itemIdx].getValue() - offset, itemIdx);
+            }
+        }
+    }
+
+    protected void updateRAfterInstantiation(int binIdx, int itemIdx) throws ContradictionException {
+        if(!R[binIdx].contains(itemIdx)) {
```

Comment:
Wouldn't be good to use the return value of `ISet#add` instead of calling `ISet#contains` + `ISet#add` ?